### PR TITLE
[v1.2] Implement Pretty Histograms

### PR DESCRIPTION
**Pull request by lucasferrari07**
- Implement printHistogram function, that uses frequencies to print a histogram of the
text letter frequency
- Change sanitize function to return only letters
- Change itemCounter function to sort letters in ASC mode

### CODE REVIEW

File: textalyze.js [hunk 1]
```diff
@@ -43,9 +43,21 @@ function stringToCharacters(text) {
...
+  return null;
+}
+
+/**
+ * Prints text frequencies histogram
+ * @param {Map} frequencies
+ */
+function printHistogram(frequencies) {
+  frequencies.forEach((freq, value) => {
+    const percentage = `[${freq < 0.1 ? ' ' : ''}${(freq * 100).toFixed(2)}%]`;
```
Comment (jfarmer) [inline textalyze.js]: Look at [String.prototype.padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart).  More broadly, any time you find yourself writing one-off code like this, double check whether there's something in the standard library that will do it (and if there is, use it).

File: textalyze.js [hunk 2]
```diff
@@ -43,9 +43,21 @@ function stringToCharacters(text) {
...
+    const length = ''.padStart(Math.floor(freq / 0.005), '=');
```
Comment (jfarmer) [inline textalyze.js]: Here I'd use `String.prototype.repeat`, so:

```javascript
const maxWidth = 200;
'='.repeat(freq * maxWidth)
```

`String.prototype.repeat` does the flooring, i.e., `str.repeat(x) === str.repeat(Math.floor(x))` for any number `x`.

File: textalyze.js (hunk 1 above)
Comment (jfarmer) [reply textalyze.js]: Er, wait, you use `padStart` below!  You could just do:

```javascript
const maxWidth = 200;
const freqStr = (freq * 100).toFixed(2).padLeft(5, ' ');
const bar = '='.repeat(freq * maxWidth);
console.log(`${value} [${freqStr}%] `${bar}`);
```

Comment (jfarmer) [reply textalyze.js]: The above code makes it absolutely clear what's going on at a glance.  That's what I'm optimizing for, here.

File: textalyze.js [hunk 3]
```diff
@@ -57,19 +69,16 @@ if (require.main === module) {
   // const printCounts = counts => console.table(counts);
   // const program = compose(printCounts, itemCounts, stringToCharacters, sanitize);
 
-  const program = x => console.table(itemCounts(stringToCharacters(sanitize(x))));
+  const program = x => printHistogram(itemFrequencyCounter(stringToCharacters(sanitize(x))));
```
Comment (jfarmer) [inline textalyze.js]: Nice!

Comment (jfarmer) [reply textalyze.js]: I wouldn't call it `program`, even those that's what I did in the example I gave.  I just wanted to show that this "program" is really a composition of several simple functions.  Give it a name like `printFileHistogram` and call `x` something more descriptive like `file` or `path` or `fileName`.

**Review by jfarmer — APPROVED** [review]

File: textalyze.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ const fs = require('fs');
  * @returns {Map} counts - A Map containing the counts of the items in the input array
  */
 function itemCounts(array) {
-  return array.reduce((map, a) => {
+  return array.sort().reduce((map, a) => {
```
Comment (jfarmer) [inline textalyze.js]: Make sure to test that this is returning things alphabetically.

Alternatively, I wouldn't rely on `itemCounts` to return a `Map` with alphabetically sorted keys.  The reason is that something could modify the returned `Map` and add keys in a non-alphabetical order.

I'd either:

1. Implement a type that acts like map but guarantees the keys are in order (like `set` vs `sortedSet` in some languages).
2. Not assume anything about the order of the data in the map and instead iterate over the keys alphabetically when it comes time to display them.

For (2), that would mean `printHistogram` looks more like:

```javascript
Array.from(frequencies.keys()).sort().foreach(key => {
  const val = frequencies.get(key);
  ...
});
```

Comment (lucasferrari07) [reply textalyze.js]: It makes a lot more sense, thanks!

File: textalyze.js (hunk 2 above)
Comment (lucasferrari07) [reply textalyze.js]: A much better way to understand, thanks!!

File: textalyze.js (hunk 1 above)
Comment (lucasferrari07) [reply textalyze.js]: Of course! I totally agree with you, thanks a lot!

File: textalyze.js [hunk 5]
```diff
@@ -43,33 +43,46 @@ function stringToCharacters(text) {
...
+    const freq = frequencies.get(letter);
+    const freqStr = (freq * 100).toFixed(2).padStart(5, ' ');
+    const bar = '='.repeat(freq * maxWidth);
+    console.log(`${letter} [${freqStr}%] ${bar}`);
+  });
 }
 
 if (require.main === module) {
-  const file = process.argv.length >= 3 ? process.argv[2] : 'sample_data/tale-of-two-cities.txt';
+  const path = process.argv.length >= 3 ? process.argv[2] : 'sample_data/tale-of-two-cities.txt';
```
Comment (jfarmer) [inline textalyze.js:66]: For a user-facing program, I wouldn't use this as a default.  Rather, I'd print a *usage* message explaining to the user how to run the program with a file and then exit with `process.exit(1)`.

File: textalyze.js [hunk 6]
```diff
@@ -43,33 +43,46 @@ function stringToCharacters(text) {
...
-  const program = x => console.table(itemCounts(stringToCharacters(sanitize(x))));
+  const printFileHistogram = (file) => {
+    printHistogram(itemFrequencyCounter(stringToCharacters(sanitize(file))));
+  };
 
-  fs.readFile(file, (err, data) => {
+  fs.readFile(path, (err, data) => {
     if (err) {
-      console.error(`Program failed opening '${file}', please try another one.`);
+      console.error(`Program failed opening '${path}', please try another one.`);
```
Comment (jfarmer) [inline textalyze.js:80]: We probably want to `process.exit(1)` here so we exit with a failed error code.

Comment (jfarmer) [reply textalyze.js:80]: Good call on `console.error` so we log to STDERR, BTW!

**Pull request merged**